engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java

```diff
@@ -64,19 +75,68 @@ public void processRecord(final TypedRecord<ProcessInstanceMigrationRecord> comm
+    while (!elementInstances.isEmpty()) {
+      final var elementInstance = elementInstances.poll();
+      final List<ElementInstance> children =
+          migrateElementInstance(elementInstance, processDefinition, mappingInstructions);
+      elementInstances.addAll(children);
+    }
+
+    stateWriter.appendFollowUpEvent(
+        processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value);
+    responseWriter.writeEventOnCommand(
+        processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value, command);
+  }
+
+  private List<ElementInstance> migrateElementInstance(
+      final ElementInstance elementInstance,
+      final DeployedProcess processDefinition,
+      final List<ProcessInstanceMigrationMappingInstructionValue> mappingInstructions) {
+
+    final String targetElementId =
+        determineTargetElementId(elementInstance, processDefinition, mappingInstructions);
+
     stateWriter.appendFollowUpEvent(
-        processInstanceKey,
+        elementInstance.getKey(),
         ProcessInstanceIntent.ELEMENT_MIGRATED,
-        processInstance
+        elementInstance
             .getValue()
             .setProcessDefinitionKey(processDefinition.getKey())
             .setBpmnProcessId(processDefinition.getBpmnProcessId())
             .setVersion(processDefinition.getVersion())
-            .setElementId(processDefinition.getBpmnProcessId()));
+            .setElementId(targetElementId));
 
-    stateWriter.appendFollowUpEvent(
-        processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value);
-    responseWriter.writeEventOnCommand(
-        processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value, command);
+    if (elementInstance.getJobKey() > 0) {
+      final var job = jobState.getJob(elementInstance.getJobKey());
+      if (job != null) {
+        stateWriter.appendFollowUpEvent(
+            elementInstance.getJobKey(),
+            JobIntent.MIGRATED,
+            job.setProcessDefinitionKey(processDefinition.getKey())
+                .setProcessDefinitionVersion(processDefinition.getVersion())
+                .setBpmnProcessId(processDefinition.getBpmnProcessId())
+                .setElementId(targetElementId));
+      }
+    }
+
+    return elementInstanceState.getChildren(elementInstance.getKey());
+  }
+
+  private static String determineTargetElementId(
+      final ElementInstance elementInstance,
+      final DeployedProcess targetProcessDefinition,
+      final List<ProcessInstanceMigrationMappingInstructionValue> mappingInstructions) {
+    if (elementInstance.getValue().getBpmnElementType() == BpmnElementType.PROCESS) {
+      // users don't provide a mapping instruction for the bpmn process id
+      return BufferUtil.bufferAsString(targetProcessDefinition.getBpmnProcessId());
+    }
+    final String elementId = elementInstance.getValue().getElementId();
+    return mappingInstructions.stream()
```

Comment:
🔧  With this method, for each child element in the process, do filter and map which means `O(m)*O(2n)` complexity (m is number of child process instances and n is the number of mapping instructions).

It would be a lot better if we create a map `sourceElementIdToTargetElementId` once and do lookups. That will have complexity of `O(n)+O(m)`. Of course, space complexity will increase to `O(n)`.

Reply:
👍 That's a great point. I'll give it a try